commit d615a0943d7181654ef74b1a7ae6f19e7a689000
Author: Phil Elwell <phil@raspberrypi.org>
Date:   Sat May 12 21:35:43 2018 +0100

    firmware/raspberrypi: Notify firmware of a reboot
    
    Register for reboot notifications, sending RPI_FIRMWARE_NOTIFY_REBOOT
    over the mailbox interface on reception.
    
    Signed-off-by: Phil Elwell <phil@raspberrypi.org>

diff --git a/drivers/firmware/raspberrypi.c b/drivers/firmware/raspberrypi.c
index 552d6067bcbf..93c3ad92097a 100644
--- a/drivers/firmware/raspberrypi.c
+++ b/drivers/firmware/raspberrypi.c
@@ -14,6 +14,7 @@
 #include <linux/module.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
+#include <linux/reboot.h>
 #include <linux/workqueue.h>
 #include <soc/bcm2835/raspberrypi-firmware.h>
 
@@ -275,6 +276,26 @@ static int rpi_firmware_get_throttled(struct rpi_firmware *fw, u32 *value)
 	return 0;
 }
 
+static int rpi_firmware_notify_reboot(struct notifier_block *nb,
+				      unsigned long action,
+				      void *data)
+{
+	struct rpi_firmware *fw;
+	struct platform_device *pdev = g_pdev;
+
+	if (!pdev)
+		return 0;
+
+	fw = platform_get_drvdata(pdev);
+	if (!fw)
+		return 0;
+
+	(void)rpi_firmware_property(fw, RPI_FIRMWARE_NOTIFY_REBOOT,
+				    0, 0);
+
+	return 0;
+}
+
 static void get_throttled_poll(struct work_struct *work)
 {
 	struct rpi_firmware *fw = container_of(work, struct rpi_firmware,
@@ -416,15 +437,32 @@ static struct platform_driver rpi_firmware_driver = {
 	.remove		= rpi_firmware_remove,
 };
 
+static struct notifier_block rpi_firmware_reboot_notifier = {
+	.notifier_call = rpi_firmware_notify_reboot,
+};
+
 static int __init rpi_firmware_init(void)
 {
-	return platform_driver_register(&rpi_firmware_driver);
+	int ret = register_reboot_notifier(&rpi_firmware_reboot_notifier);
+	if (ret)
+		goto out1;
+	ret = platform_driver_register(&rpi_firmware_driver);
+	if (ret)
+		goto out2;
+
+	return 0;
+
+out2:
+	unregister_reboot_notifier(&rpi_firmware_reboot_notifier);
+out1:
+	return ret;
 }
 subsys_initcall(rpi_firmware_init);
 
 static void __init rpi_firmware_exit(void)
 {
 	platform_driver_unregister(&rpi_firmware_driver);
+	unregister_reboot_notifier(&rpi_firmware_reboot_notifier);
 }
 module_exit(rpi_firmware_exit);
 
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1251_d615a0943d7181654ef74b1a7ae6f19e7a689000.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1251_d615a0943d7181654ef74b1a7ae6f19e7a689000.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
